src/content/advanced/connectivity/ingress/multi-nginx-ic/index.md

```diff
@@ -79,13 +79,12 @@ This is how one can achieve it by using multiple Ingress NGINX Controllers:
 
     ```yaml
     controller:
-      ingressClass: nginx-internal
       ingressClassResource:
         name: nginx-internal
         controllerValue: k8s.io/ingress-nginx-internal
       service:
         public: false
-        subdomain: "*.ingress-internal"
```

Comment:
The public Ingress NGINX Controller by default uses `ingress.<domain>.<tld>` and vintage clusters reference to this by a wildcard entry in the hosted zone.

This wildcard entry might not exist for some CAPI flavors, so you then could use `*.<domain>.<tld>` for the public Ingress NGINX Controller service.

Even though this still is a ugly solution (I've been managing around 600 public zones for a while and saw a lot of different use cases, trust me :D), it would not be easy to understand and logic to customers that the internal one then is called `*.ingress-internal.<domain>.<tld>`.

So you either have...

- ... a host record `ingress.<domain>.<tld>` pointing to the public load balancer service's IP address, ...
- ... another host record `ingress-internal.<domain>.<tld>` pointing to the internal load balancer service's IP address...
- ... and customers using the pre-created wildcard record or creating dedicated records / one wildcard record pointing to the according host record...

... or ...

- ... a host record `*.<domain>.<tld>` pointing to the public load balancer service's IP address, ...
- ... another host record `*.internal.<domain>.<tld>` pointing to the internal load balancer service's IP address...
- ... and customers using this wildcard record.

Also keep in mind that our default vintage cluster domains are already close to FQDN length limit for certificates. I'd bet `*.ingress-internal.<domain>.<tld>` will fail on issuing certificates using Let's Encrypt.

At best customers are using External DNS to create single host records for their Ingress resources instead of using a wildcard anywhere, especially because it's hard to debug and a NXDOMAIN still is better than traffic being sent to Ingress NGINX Controller by mistake.

Reply:
> The public Ingress NGINX Controller by default uses ingress.<domain>.<tld> and vintage clusters reference to this by a wildcard entry in the hosted zone.

That's clear.

> This wildcard entry might not exist for some CAPI flavors, so you then could use *.<domain>.<tld> for the public Ingress NGINX Controller service.

That's a CAPI problem. From the product perspective we expect the wildcard to be there. If that not possible for X reason, it needs to be documented for the specific provider.

> it would not be easy to understand and logical to customers that the internal one then is called *.ingress-internal.<domain>.<tld>.

Maybe switching to `*.internal.<baseDomain>` would be better. sure.

> So you either have...
>    ... a host record ingress.<domain>.<tld> pointing to the public load balancer service's IP address, ...
>    ... another host record ingress-internal.<domain>.<tld> pointing to the internal load balancer service's IP address...
>    ... and customers using the pre-created wildcard record or creating dedicated records / one wildcard record pointing to the according host record...
>
> ... or ...
>
>   ... a host record *.<domain>.<tld> pointing to the public load balancer service's IP address, ...
>   ... another host record *.internal.<domain>.<tld> pointing to the internal load balancer service's IP address...
>   ... and customers using this wildcard record.

neither is true.

Reality is that in order to meet the product criteria `Ingress working out of the box` for internal ingresses is that you need to attach the wildcard to the IC service. Then you can use your ingresses like `<whatever>.internal.<baseDomain>`.

> Also keep in mind that our default vintage cluster domains are already close to FQDN length limit for certificates. I'd bet *.ingress-internal.<domain>.<tld> will fail on issuing certificates using Let's Encrypt.

That's true, maybe `shortening the name to `*.internal.<domain>`?

> At best customers are using External DNS to create single host records for their Ingress resources instead of using a wildcard anywhere, especially because it's hard to debug and a NXDOMAIN still is better than traffic being sent to Ingress NGINX Controller by mistake.

That's an assumption we should not make.

Reply:
I don't feel like the whole multiple Ingress Controllers thing is in the scope of `Ingress working out of the box`. This is fulfilled by the default settings and the above is already not "out of the box" as you need to customize it. So therefore I'd rather go for the solution which is more in alignment with the existing Ingress Controller in the cluster than adding a second Ingress Controller with a complete different configuration as this - again - makes trouble shooting more difficult and the whole setup harder to understand for customers.

By default Ingress NGINX Controller creates a `ingress.<domain>.<tld>` record. The internal service, if enabled via `controller.service.internal.enabled`, creates `ingress-internal.<domain>.<tld>` by default (default means: no additional values set, just click install).

The whole "there's already a wildcard in the zone pointing to `ingress.<domain>.<tld>` is not in the scope of the Ingress NGINX Controller and I therefore do not think it's its responsibility to create this record.

This is why I'd rather keep it simple, understandable and aligned with the normal public Ingress NGINX Controller.

Maybe customers have a completely internal domain not reachable via internet with a wildcard pointing to `ingress-internal.<domain>.<tld>`, who knows. At least I'd not ask customers to add `*.ingress-internal.<domain>.<tld>` here as this might feel like a best practice even though it does not apply to their use case.

Reply:
Additionally: I never saw a single customer really using this mechanism for production workload. Most of them are creating CNAME records or A records in different zones manually which then point to either the `ingress.<domain>.<tld>`, `ingress-internal.<domain>.<tld>` or cloud provider load balancer's host record.

If we tell them to use `*.ingress-internal.<domain>.<tld>` here, they can not create CNAME records pointing to `ingress-internal.<domain>.<tld>` as this would not be created then.

We are talking about internal zones and from a security point of view it is not recommended to publish those records to the internet at all, which brings us back to using internal-only zones with CNAMEs pointing to `ingress-internal.<domain>.<tld>` instead of using it directly.

Reply:
I think the point here is to whom this guide is directed to. `Ingress working out of the box` with a few configs here is not the same than needing to configure nginx+external-dns+cert-manager just to make this scenario work.
Customers that are onboarding to our platform might find that too much to do for an internal zone. Customers with existing infrastructure and years of experience won't even read this guide probably.

Maybe @LolloneS can give some insights as he was the original author of the change and has a better understanding of what customers or potential customers are doing.

Reply:
I'll be honest, I don't know much about the technical details here. What I know is that it took me and Mati an hour to set up a demo hello-world app in a CAPZ cluster because a lot of stuff was missing out of the box, and if I understand correctly the wildcard was part of it. This also broke the _public_ ingress (which was the only one being deployed at the time).

In the end, if we want to be a developer platform, this stuff should be easy. If the wildcard is not the solution, then IMHO first we work on an automated solution to replace the wildcard, and only then we remove the wildcard. We don't just remove the wildcard and leave customers wondering why a demo app doesn't work. If some assumption doesn't work, I believe we should clearly document it.

Btw `ingressClass: nginx-internal` is needed, otherwise if you look at the nginx deployment it runs with `--ingress-class=nginx` under the hood instead of using `nginx-internal`. From `helm/ingress-nginx/templates/_params.tpl`: `- --ingress-class={{ .Values.controller.ingressClass }}`

Reply:
`controller.ingressClass` is deprecated behavior as it only sets the ingress class name to look for in annotations. The new behavior is looking for IngressClasses with `controller` matching the `--controller-value` flag of the Ingress NGINX Controller and then looking for Ingresses with `ingressClassName` matching the `name` of the IngressClass.

Reply:
Also remember: This particular document is not about publishing a demo app, it's about running multiple Ingress Controllers side-by-side in production. In some cases the second one is not installed for internal traffic but for other stuff, so I would not frame those docs into "you just started your cluster, here you go" but into "extended use case, change this to your needs".